gh-133296: Publicly expose critical section API that accepts PyMutex (#135899)
python/cpython: Doc/c-api/init.rst

provides weaker guarantees than traditional locks -- they are useful because
their behavior is similar to the :term:`GIL`.

> Variants that accept :c:type:`PyMutex` instances rather than objects are also

Reviewer: ```suggestion
Variants that accept :c:type:`PyMutex` pointers rather than Python objects are also
```

   .. versionadded:: 3.13

.. c:macro:: Py_BEGIN_CRITICAL_SECTION_MUTEX(m)

   Locks the mutex *m* and begins a critical section.

   In the free-threaded build, this macro expands to::

     {
          PyCriticalSection _py_cs;
          PyCriticalSection_BeginMutex(&_py_cs, m)

   Note that unlike :c:macro:`Py_BEGIN_CRITICAL_SECTION`, there is no cast for
   the second argument - it must be a :c:type:`PyMutex`.

   On the default build, this macro expands to ``{``.

>    .. versionadded:: 3.15

Reviewer: Per @vstinner, I think we use `versionadded:: next` here but I'm not entirely sure why.

```suggestion
   .. versionadded:: next
```

Reviewer 2: Because it will show up as "unreleased" on the docs until a1.

   .. versionadded:: 3.13

.. c:macro:: Py_BEGIN_CRITICAL_SECTION2_MUTEX(m1, m2)

   Locks the mutexes *m1* and *m2* and begins a critical section.

   In the free-threaded build, this macro expands to::

     {
          PyCriticalSection2 _py_cs2;
          PyCriticalSection2_BeginMutex(&_py_cs2, m1, m2)

   Note that unlike :c:macro:`Py_BEGIN_CRITICAL_SECTION2`, there is no cast for

>    the second and third arguments - they must be :c:type:`PyMutex` pointers.

Reviewer 3: the macro has only two arguments, this should be first and second arguments

Author: thanks! Also fixed the same issue in the other doc section.

>    the second and third arguments - they must be :c:type:`PyMutex` instances.

Reviewer: ```suggestion
   the second and third arguments - they must be :c:type:`PyMutex` pointers.
```

   On the default build, this macro expands to ``{``.

>    .. versionadded:: 3.15

Reviewer: ```suggestion
   .. versionadded:: next
```
